Sources/CommonMark/Visitor.swift

```diff
@@ -0,0 +1,387 @@
+public enum VisitorContinueKind {
+    /// The visitor should visit the descendents of the current node.
+    case visitChildren
+
+    /// The visitor should avoid visiting the descendents of the current node.
+    case skipChildren
+
+    case inherit
+
+    /// The default is `.visitChildren`
+    static let `default`: Self = .visitChildren
+
+    mutating func override(with other: Self) {
+        switch other {
+        case .visitChildren, .skipChildren:
+            self = other
+        case .inherit:
+            break
+        }
+    }
+}
+
+/// Visitor for walking a visitable structure.
+///
+/// The order of object-wise visitations is:
+///
+/// 1. `func visit(node:)`, iff the visited object conforms to `Node`.
+/// 2. `func visit(block:)`, iff the visited object conforms to `Block`.
+/// 3. `func visit(containerBlock:)`, iff the visited object conforms to `ContainerBlock`.
+/// 4. `func visit(leafBlock:)`, iff the visited object conforms to `LeafBlock`.
+/// 5. `func visit(inline:)`, iff the visited object conforms to `Inline`.
+/// 6. `func visit(<concrete>:)`, where `<concrete>` corresponds to the
+///   the visited object's concrete type.
+///
+/// With each visitation's returned `VisitorContinueKind` overriding the previous one.
+///
+public protocol Visitor {
+    /// The fallback for when `.inherit`
+    var defaultContinueKind: VisitorContinueKind { get }
```

Comment:
Do we need this if we have `visit(node:)`?

Reply:
Nope, fixed.